### PR TITLE
Update freedesktop runtime to 23.08

### LINKED ISSUE #4: runtime org.freedesktop.Platform branch 20.08 is end-of-life
org.freedesktop.Platform 20.08 is no longer receiving fixes and security updates. Please update to a supported runtime version.

### PR DESCRIPTION
**Pull request by takluyver**
Closes #4.

### CODE REVIEW

**Comment by takluyver** [comment]
bot, build

**Comment by takluyver** [comment]
Hmm, looks like there's some problem with the Flathub infrastructure. I'll try again later.

**Comment by hfiguiere** [comment]
bot, build

**Comment by takluyver** [comment]
Now stuck on https://github.com/flathub/shared-modules/issues/201

**Comment by bbhtt** [comment]
@takluyver Can you try now? Pygame issue should be fixed https://github.com/flathub/shared-modules/commit/b5eb7c51ed89abee30e703f5037d92abd2ceee53

Update the shared-modules.

File: com.inventwithpython.flippy.json [hunk 1]
```diff
@@ -2,15 +2,15 @@
     "id": "com.inventwithpython.flippy",
     "runtime": "org.freedesktop.Platform",
     "sdk": "org.freedesktop.Sdk",
-    "runtime-version": "20.08",
+    "runtime-version": "22.08",
```
Comment (bbhtt) [inline com.inventwithpython.flippy.json]: Can you try 23.08?

**Comment by bbhtt** [comment]
This is using python2. You need to build cython with python2 separately see https://github.com/flathub/shared-modules/issues/201#issuecomment-1715574388

File: com.inventwithpython.flippy.json [hunk 2]
```diff
@@ -2,15 +2,15 @@
     "id": "com.inventwithpython.flippy",
     "runtime": "org.freedesktop.Platform",
     "sdk": "org.freedesktop.Sdk",
-    "runtime-version": "20.08",
+    "runtime-version": "23.08",
     "command": "/app/bin/flippy",
     "finish-args": [
        "--socket=x11",
```
Comment (bbhtt) [inline com.inventwithpython.flippy.json:8]: Add `--device=dri`, this is a graphical application.

File: com.inventwithpython.flippy.json [hunk 3]
```diff
@@ -2,15 +2,26 @@
     "id": "com.inventwithpython.flippy",
     "runtime": "org.freedesktop.Platform",
     "sdk": "org.freedesktop.Sdk",
-    "runtime-version": "20.08",
+    "runtime-version": "23.08",
     "command": "/app/bin/flippy",
```
Comment (bbhtt) [inline com.inventwithpython.flippy.json]: Command should not be the absolute path, just the name of the executable

**Comment by bbhtt** [comment]
Why is python-2.7 being installed while launch.py uses python3? https://github.com/flathub/com.inventwithpython.flippy/blob/03898ca97023ed73a32b72758727f5e4e183438f/launch.py#L1C1-L1C19

**Comment by bbhtt** [comment]
> Why is python-2.7 being installed while launch.py uses python3? https://github.com/flathub/com.inventwithpython.flippy/blob/03898ca97023ed73a32b72758727f5e4e183438f/launch.py#L1C1-L1C19

Probably the reason of the run faliure:

```
[📦 com.inventwithpython.flippy ~]$ python3
Python 3.11.5 (main, Nov 10 2011, 15:00:00) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pygame.locals import *
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'pygame'

[📦 com.inventwithpython.flippy ~]$ python2
Python 2.7.18 (default, Sep 12 2023, 16:12:55) 
[GCC 13.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from pygame.locals import *
pygame 1.9.6
Hello from the pygame community. https://www.pygame.org/contribute.html
>>>
```

**Comment by bbhtt** [comment]
> Probably the reason of the run faliure:

Works after changing to python2. Which python version does it actually need? pygame works on both...

```
$ flatpak run com.inventwithpython.flippy 

pygame 1.9.6
Hello from the pygame community. https://www.pygame.org/contribute.html
ALSA lib ../../pulse/pulse.c:242:(pulse_connect) PulseAudio: Unable to connect: Connection refused
```

Did sound used to work? Sounds like it could use `--socket=pulseaudio`.

**Comment by takluyver** [comment]
Yeah, I must have got confused somewhere along the years. I think it was using Python 2 when I first packaged it up, but it looks like I partly changed to Python 3 somewhere along the line. It's meant to be an example, so I'll have a go at making it use Python 3.

**Comment by bbhtt** [comment]
It was working all these years because pygame was built using `pip3` which explicitly is `python3` :P 

I switched it to `python` so that when multiple python is installed in `/app` and runtime, `python` will point to whatever version installed in `/app` first, then to the one in runtime. (as `/app` comes first in path over runtime)

**Comment by takluyver** [comment]
> Did sound used to work? Sounds like it could use --socket=pulseaudio.

There is no sound. I'd guess that it calls something in pygame that sets up audio, but it's never actually used. It's a very simple game used [as an example in a programming book](https://inventwithpython.com/pygame/chapter10.html).

> Add --device=dri, this is a graphical application.

My very loose understanding was that DRI was something you only need if you have fancy graphics like OpenGL, which this doesn't. Is that right?

> Command should not be the absolute path, just the name of the executable

OK, I can do that.

**Comment by bbhtt** [comment]
> My very loose understanding was that DRI was something you only need if you have fancy graphics like OpenGL, which this doesn't. Is that right?

Yea this seems to not use it at all.

File: com.inventwithpython.flippy.json (hunk 3 above)
Comment (bbhtt) [reply com.inventwithpython.flippy.json]: Added in https://github.com/flathub/com.inventwithpython.flippy/pull/7/commits/58f8b0d101da4c6d747055a0b4bb0078863f0320

File: com.inventwithpython.flippy.json (hunk 2 above)
Comment (bbhtt) [reply com.inventwithpython.flippy.json:8]: Not required.

File: com.inventwithpython.flippy.json (hunk 1 above)
Comment (bbhtt) [reply com.inventwithpython.flippy.json]: Done in https://github.com/flathub/com.inventwithpython.flippy/pull/7/commits/238485aa6365949f0607a62f9854cf5b0d96bc8f

**Comment by takluyver** [comment]
Is there still a convenient way of testing the app built from the PR directly? IIRC at one point the bot gave you a download link, but either that's no longer the case or I've misremembered.

**Comment by takluyver** [comment]
... or I just needed more patience. :laughing:

**Comment by takluyver** [comment]
I've tested it now, it seems to work. :+1: Thanks for all your help with figuring out the various issues!

**Pull request merged**